File: src/main/java/org/springframework/samples/petclinic/rest/mapper/PetMapper.java
package org.springframework.samples.petclinic.rest.mapper;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.springframework.samples.petclinic.owner.Pet;
import org.springframework.samples.petclinic.rest.dto.PetDto;

@Mapper(componentModel = "spring", uses = { VisitMapper.class, PetTypeMapper.class })
public interface PetMapper {

	@Mapping(source = "type", target = "type")
	PetDto toDto(Pet pet);

}

File: src/main/kotlin/org/springframework/samples/petclinic/rest/mapper/PetMapper.kt
package org.springframework.samples.petclinic.rest.mapper

import org.springframework.samples.petclinic.owner.Pet
import org.springframework.samples.petclinic.rest.dto.PetDto
import org.springframework.stereotype.Component

@Component
class PetMapper(
    private val visitMapper: VisitMapper,
    private val petTypeMapper: PetTypeMapper
) {

    fun toDto(pet: Pet): PetDto {
        return PetDto(
            id = pet.id,
            name = pet.name,
            birthDate = pet.birthDate,
            type = pet.type?.let { petTypeMapper.toDto(it) },
            visits = pet.getVisits().map { visitMapper.toDto(it) }
        )
    }

}